rust/worker/src/segment/distributed_hnsw_segment.rs

```diff
@@ -172,57 +173,57 @@ impl DistributedHNSWSegmentWriter {
+                            }
+                        },
+                    };
+                    self.index.read().add(record.offset_id as usize, embedding);
                 }
+                // This shouldn't be reached since materialization always derefs
+                // upserts into either updates or inserts.
                 Operation::Upsert => {
-                    // hnsw index behavior is to treat add() as upsert
-                    let segment_offset_id = record.0.segment_offset_id;
-                    // Assumption: Upserts must have embedding set
-                    let embedding = record.0.log_record.record.embedding.as_ref().unwrap();
-                    self.index
-                        .read()
-                        .add(segment_offset_id as usize, &embedding);
+                    panic!(
+                        "Invariant violation. Upserts should not be present after materialization"
+                    );
                 }
                 Operation::Update => {
-                    // hnsw index behvaior is to treat add() as upsert so this
-                    // will update the embedding
-                    // The assumption is that materialized log records only contain
-                    // valid updates.
-                    let segment_offset_id = record.0.segment_offset_id;
-                    match record.0.log_record.record.embedding.as_ref() {
-                        Some(e) => {
-                            self.index.read().add(segment_offset_id as usize, &e);
-                        }
-                        None => {
-                            // An update may not necessarily update the embedding
-                            continue;
-                        }
+                    // Should panic here if embedding is not found because it likely
+                    // means that somehow our storage is corrupt as data record on
+                    // the record segment does not contain the embedding.
+                    let embedding = match record.final_embedding {
+                        Some(e) => e,
+                        None => match record.data_record.as_ref() {
+                            Some(record) => record.embedding,
+                            None => {
+                                panic!("Invariant violation. Embedding not found on storage");
+                            }
+                        },
                     };
+                    self.index.read().add(record.offset_id as usize, embedding);
                 }
                 Operation::Delete => {
-                    // The assumption is that materialized log records only contain
-                    // valid deletes
-                    let segment_offset_id = record.0.segment_offset_id;
-                    self.index.read().delete(segment_offset_id as usize);
+                    self.index.read().delete(record.offset_id as usize);
```

Comment:
same nit about the comment being removed

Reply:
The comment `the assumption is that materialized log records only contain valid deletes` is no longer valid because the materializer explicitly has verified that all the deletes pertain to records that actually exist in storage

Reply:
I think that is what the comment is intended to reflect? That the offset id will always be valid

Reply:
ah I see what you mean